activation/poetlistener.go

```diff
@@ -0,0 +1,74 @@
+package activation
+
+import (
+	"bytes"
+	xdr "github.com/nullstyle/go-xdr/xdr3"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
+	"github.com/spacemeshos/go-spacemesh/types"
+)
+
+const PoetProofProtocol = "PoetProof"
+
+type PoetListener struct {
+	Log               log.Log
+	net               service.Service
+	poetDb            *PoetDb
+	poetProofMessages chan service.GossipMessage
+	started           bool
+	exit              chan struct{}
+}
+
+func (l *PoetListener) Start() {
+	if l.started {
+		return
+	}
+	go l.loop()
+	l.started = true
+}
+
+func (l *PoetListener) Close() {
+	close(l.exit)
+	l.started = false
+}
+
+type poetProofMessage struct {
+	types.PoetProof
+	PoetId    []byte
+	RoundId   uint64
+	Signature []byte
+}
+
+func (l *PoetListener) loop() {
+	for {
+		select {
+		case poetProof := <-l.poetProofMessages:
+			var proofMessage poetProofMessage
```

Comment:
notice that you have only one reader on this chan and that reader blocks the gossip channel while it dose some processing and access the database, you should do it in a goroutine. blocking a gossip chan can stall p2p if the buffer get filled.

Reply:
great catch!